src/icd/icdbuilder.cpp

```diff
@@ -0,0 +1,170 @@
+{
+    return line.startsWith(GENERATED_ICD_PREFIX);
+}
+
+BuildStep *modifiedMakeBuildStep(BuildStepList *steps)
+{
+    auto step = steps->firstOfType<MakeStep>();
+    if (!step)
+        return nullptr;
+    step->setBuildTarget(ICD_TARGET_NAME, true);
+    return step;
+}
+
+BuildStep *modifiedCMakeBuildStep(BuildStepList *steps)
+{
+    for (auto step : steps->steps()) {
+        auto map = step->toMap();
+        if (map.contains(CMAKE_TARGETS_KEY)) {
+            map[CMAKE_TARGETS_KEY] = ICD_TARGET_NAME;
+            step->fromMap(map);
+            return step;
+        }
+    }
+    return nullptr;
+}
+
+BuildStep *modifiedBuildStep(BuildStepList *steps)
+{
+    auto makeStep = modifiedMakeBuildStep(steps);
+    if (makeStep != nullptr)
+        return makeStep;
+    return modifiedCMakeBuildStep(steps);
+}
+
+std::shared_ptr<BuildConfiguration> prepareClone(BuildConfiguration *buildConfig)
+{
+    auto result = std::shared_ptr<BuildConfiguration>{buildConfig};
+    QObject::connect(BuildManager::instance(),
+                     &BuildManager::buildQueueFinished,
+                     [result](bool) mutable { result.reset(); });
+    return result;
+}
+
+std::shared_ptr<BuildConfiguration> config(Project *project)
+{
+    const auto target = project->activeTarget();
+    if (!target)
+        return nullptr;
+
+    const auto buildConfig = target->activeBuildConfiguration();
+    const auto factory = IBuildConfigurationFactory::find(target);
+
+    if (!factory || !buildConfig)
+        return nullptr;
+
+    return prepareClone(factory->clone(target, buildConfig));
+}
+
+BuildStepList *buildSteps(Project *project)
+{
+    auto buildConfig = config(project);
+    if (buildConfig == nullptr)
+        return nullptr;
+    return buildConfig->stepList(Core::Id(ProjectExplorer::Constants::BUILDSTEPS_BUILD));
+}
+
+void displayWarning()
+{
+    QMessageBox::warning(Core::ICore::dialogParent(),
+                         Asn1AcnPlugin::tr("Build ICD failed"),
+                         Asn1AcnPlugin::tr(
+                             "Incorrect project configuration - command supported only in "
+                             "qmake and cmake project created by plugin."),
+                         QMessageBox::Ok);
+}
+
+void addStepOutputListener(BuildStep *step)
+{
+    QObject::connect(step,
+                     &BuildStep::addOutput,
+                     [](const QString &string,
+                        BuildStep::OutputFormat,
+                        BuildStep::OutputNewlineSetting) {
+                         if (icdGenerationFinished(string))
+                             openGeneratedIcd(string);
+                     });
+}
+
+} // namespace
+
+IcdBuilder::IcdBuilder() {}
+IcdBuilder::~IcdBuilder() {}
+
+void IcdBuilder::run(ProjectExplorer::Project *project)
+{
+    auto steps = buildSteps(project);
+    auto step = (steps != nullptr) ? modifiedBuildStep(steps) : nullptr;
+
+    if (!step) {
```

Comment:
I believe != nullptr is the preferred idiom

Reply:
too many years using smart ptrs without C++11 :) thanks, fixed